src/Compilers/Core/MSBuildTask/ManagedToolTask.cs

```diff
@@ -0,0 +1,97 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using Microsoft.Build.Utilities;
+
+namespace Microsoft.CodeAnalysis.BuildTasks
+{
+    public abstract class ManagedToolTask : ToolTask
+    {
+        protected abstract bool IsManagedTask { get; }
+
+        protected abstract string CommandLineArguments { get; }
```

Comment:
At first, having both `CommandLineArguments` and `GenerateCommandLineCommands` confused me since it seems like they're the same thing. However, looking at the code it seems like they're different.

Could you let me know if my reading is correct? `CommandLineArguments` in this context means the set of arguments that are intended to be passed to the tool. This is irrespective of calling convention (whether the tool is invoked directly or through a managed host). `GenerateCommandLineCommands`, however, is strictly about generating a list of arguments to be passed to CreateProcess -- there's no distinguishing between arguments to the tool and arguments to the host.

If my prior read is correct, I think this is a useful distinction, but I would rename CommandLineArguments something like `ToolArguments` and add a doc comment to make the relationship between `ToolArguments` and `GenerateCommandLineCommands()` clear.

Reply:
Correct.

`GenerateCommandLineCommands` is called by MSBuild, and it expects the arguments to the actual process invocation.

The corresponding wrapper for that, which is the clean "just the actual arguments to the tool", is `CommandLineArguments` (now renamed to `ToolArguments` and doc'd)